src/mainboard/pcengines/apu1/mainboard.c

```diff
@@ -316,32 +317,42 @@ const char *smbios_mainboard_serial_number(void)
 {
 	static char serial[10];
 	struct device *dev;
-	uintptr_t bar18;
 	u32 mac_addr = 0;
+	u32 val32 = 0;
 	int i;
 
 	/* Already initialized. */
 	if (serial[0] != 0)
 		return serial;
 
 	dev = pcidev_on_root(4, 0);
+
 	if (dev)
 		dev = pcidev_path_behind(dev->link_list, PCI_DEVFN(0, 0));
 	if (!dev)
 		return serial;
 
+	/* Read PCI Configuration Space Bar 2 (0x18 offset).
+	 * Bits 31-4 contain memory address where internal NIC registers are mapped.
+	 * val32 stores that address.
+	 */
+	val32 = pci_io_read_config32(PCI_DEV(1, 0, 0), PCI_BAR_2_REG);
+	val32 &= 0xFFFFFFF0;
+
+	u8* mac_reg_addr = (u8 *)val32;
+
 	/* Read in the last 3 bytes of NIC's MAC address. */
-	bar18 = pci_read_config32(dev, 0x18);
-	bar18 &= 0xFFFFFC00;
```

Comment:
It seems like everything was well coded except the BAR mask. Am I wrong?

Reply:
As I mentioned in the other comment, the problem is actually faulty read via `pci_read_config32(dev, 0x18);`. The BAR mask is rather minor error.